app/controllers/users/invitations_controller.rb

```diff
@@ -1,68 +1,75 @@
 
-    organisation = Organisation.find(organisation_id)
-    user = User.find_by(email: invite_params[:email])
-    if user.organisations.empty?
-      user.organisations << organisation
+    if user_belongs_to_our_organisation?(organisation)
+      self.resource = invite_resource
+      render :new and return
     end
+
+    invite_user(organisation)
+
+    redirect_to(after_path(organisation), notice: "#{invited_user.email} has been invited to join #{organisation.name}") and return
+  end
+
+private
+
+  def invite_user(organisation)
+    membership = invited_user.memberships.find_or_create_by(invited_by_id: current_user.id, organisation: organisation)
+    membership.update_attributes(can_manage_team: params[:can_manage_team], can_manage_locations: params[:can_manage_locations])
+
+    send_invitation_email(membership)
   end
 
-  def create_membership
-    token = Devise.friendly_token[0, 20]
-    invited_user.memberships.create!(
-      invited_by_id: current_user.id,
-      organisation: current_organisation,
-      invitation_token: token
-    )
+  def send_invitation_email(membership)
+    AuthenticationMailer.membership_instructions(invited_user, membership.invitation_token, organisation: current_organisation).deliver_now
+  end
 
-    AuthenticationMailer.membership_instructions(invited_user, token, organisation: current_organisation).deliver_now
+  def after_path(organisation)
+    super_admin? ? admin_organisation_path(organisation) : created_invite_memberships_path
+  end
 
-    redirect_to team_members_path, notice: "#{invited_user.email} has been invited to join #{current_organisation.name}"
+  def user_belongs_to_our_organisation?(organisation)
+    invited_user.confirmed? && invited_user.organisations.include?(organisation)
   end
 
   def user_belongs_to_other_organisations?
     invited_user.present? &&
-      invited_user.organisations.present? &&
+      invited_user.confirmed? &&
       invited_user.organisations.pluck(:id).exclude?(current_organisation.id)
   end
 
   def authenticate_inviter!
     # https://github.com/scambra/devise_invitable#controller-filter
-    redirect_to(root_path) unless current_user&.can_manage_team?
+    redirect_to(root_path) and return unless current_user&.can_manage_team?(current_organisation)
   end
 
   def delete_user_record
     invited_user.destroy!
   end
 
-  def return_user_to_invite_page
-    respond_with_navigational(resource) { render :new }
-  end
-
   def set_target_organisation
     @target_organisation = Organisation.find(params[:organisation_id])
   end
 
   def user_is_invalid?
-    # This is an indirect solution to preventing a user being re-invited when they belong
-    # to another organisation.
-    self.resource = resource_class.new(invite_params)
-    resource.invalid?
+    invite_params[:email].match? Devise.email_regexp
   end
 
   def invited_user
-    @invited_user ||= User.find_by(email: invite_params[:email])
+    User.find_by(email: invite_params[:email])
```

Comment:
Memoization caused a bug, but not having it will result in a large number of db queries.
Need to find a middleground.